qiskit_experiments/library/tomography/fitters/cvxpy_lstsq.py

```diff
@@ -157,123 +166,173 @@ def cvxpy_linear_lstsq(
-    metadata["cvxpy_status"] = []
-    for arg, con in zip(args, cons):
-        # Optimization problem
-        obj = cvxpy.Minimize(cvxpy.norm(arg, p=2))
-        prob = cvxpy.Problem(obj, con)
-
-        # Solve SDP
-        cvxpy_utils.solve_iteratively(prob, 5000, **kwargs)
-
-        # Return optimal values and problem metadata
-        metadata["cvxpy_solver"] = prob.solver_stats.solver_name
-        metadata["cvxpy_status"].append(prob.status)
-
+        num_meas_cond = 0
+        cond_meas_data = np.zeros((measurement_data.shape[0], 0), dtype=int)
+        cond_meas_indices = np.zeros((1, 0), dtype=int)
+
+    if conditional_preparation_indices:
+        cond_prep_data = preparation_data[:, np.array(conditional_preparation_indices, dtype=int)]
+        cond_prep_indices = np.unique(cond_prep_data, axis=0)
+        num_prep_cond = len(cond_meas_indices)
+        metadata["conditional_preparation_index"] = []
+    else:
+        num_prep_cond = 0
+        cond_prep_data = np.zeros((preparation_data.shape[0], 0), dtype=int)
+        cond_prep_indices = np.zeros((1, 0), dtype=int)
+
+    if outcome_data.shape[0] > 1:
+        metadata["conditional_circuit_outcome"] = []
+
+    fits = []
+    for cond_prep_idx in cond_prep_indices:
+        for cond_meas_idx in cond_meas_indices:
+            # Mask for specified conditional indices
+            cond_mask = np.all(cond_meas_data == cond_meas_idx, axis=1) & np.all(
+                cond_prep_data == cond_prep_idx, axis=1
+            )
+            if weights is None:
+                cond_weights = None
+            else:
+                cond_weights = weights[:, cond_mask]
+
+            basis_matrix, probability_data, probability_weights = lstsq_utils.lstsq_data(
+                outcome_data[:, cond_mask],
+                shot_data[cond_mask],
+                measurement_data[cond_mask],
+                preparation_data[cond_mask],
+                measurement_basis=measurement_basis,
+                preparation_basis=preparation_basis,
+                measurement_qubits=measurement_qubits,
+                preparation_qubits=preparation_qubits,
+                weights=cond_weights,
+                conditional_measurement_indices=conditional_measurement_indices,
+                conditional_preparation_indices=conditional_preparation_indices,
+            )
+
+            # Since CVXPY only works with real variables we must specify the real
+            # and imaginary parts of matrices seperately: rho = rho_r + 1j * rho_i
+
+            num_circ_components, num_tomo_components, _ = probability_data.shape
+            dim = int(np.sqrt(basis_matrix.shape[1]))
+
+            # Generate list of conditional components for block diagonal matrix
+            # rho = sum_k |k><k| \otimes rho(k)
+            rhos_r = []
+            rhos_i = []
+            cons = []
+            for i in range(num_circ_components):
+                for j in range(num_tomo_components):
```

Comment:
What is the expected size of these elements? Quadruple loop seems quite heavy, but of course it depends on the element size.

Reply:
It is what it is. It of course depends on the tomography you are doing. You are basically trading the dimension of the optimization problem for the number of components. eg for 3 qubits, instead of 1 8x8 matrix, you can have 2 4x4 matrices, or 4 2x2 or 8 1x1 etc

Reply:
ah okey, the outer loops are on conditional elements so there is actually no loop without conditioning.